express-notes-api/index.js

```diff
@@ -0,0 +1,114 @@
+const express = require('express');
+const fs = require('fs');
+
+const app = express();
+const expressJson = express.json();
+
+app.listen(3000, () => {
+  // eslint-disable-next-line no-console
+  console.log('Express listening on port 3000.');
+});
+
+app.use(expressJson);
+
+let data;
+fs.readFile('data.json', 'utf-8', (err, datafile) => {
+  if (err) console.error(err);
+  data = JSON.parse(datafile);
+});
+
+app.get('/api/notes', (req, res) => {
+  res.status(200).json(Object.values(data.notes));
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  const id = req.params.id;
+
+  if (!Number.isInteger(Number(id))) {
+    res.status(400).json({ error: 'id must be a positive integer' });
+    return;
+  }
+  if (!Object.keys(data.notes).includes(id)) {
+    res.status(404).json({ error: `cannot find note with id ${id}` });
+    return;
+  }
```

Comment:
This could be more efficient. Since you already have the key, you can use it to check for the existence of the object directly 

```js
if (!data.notes[id]) {
```

Look up "constant time" vs "linear time"

Reply:
ohhh gotcha. includes would involve iterating, whilst simply searching within the object for the property if it exists would be much quicker and take advantage of the fact that the notes are stored as an object.